stockflux-launcher/src/free-text-search/FreeTextSearch.js

```diff
@@ -0,0 +1,216 @@
+    resultsWindow = newReference;
+  };
+
+  const resetLauncherInputField = () => {
+    if (searchInputRef.current) searchInputRef.current.value = '';
+  };
+
+  const closeResultsWindow = useCallback(() => {
+    if (resultsWindow) {
+      resultsWindow.close();
+      setResultsWindow(null);
+    }
+  }, []);
+
+  const handleOnInputChange = useCallback(
+    event => {
+      setQuery(event.target.value);
+      if (event.target.value !== null && event.target.value.length === 0) {
+        closeResultsWindow();
+      } else if (!resultsWindow)
+        createWindow(
+          searchButtonRef,
+          searchInputRef,
+          props.dockedTo,
+          windowState.bounds,
+          setResultsWindow
+        ).catch(err => console.error(err));
+    },
+    [closeResultsWindow, props.dockedTo, windowState.bounds]
+  );
+
+  const handleSearchClick = useCallback(async () => {
+    if (resultsWindow) {
+      closeResultsWindow();
+      if (props.dockedTo === Constants.ScreenEdge.TOP)
+        resetLauncherInputField();
+    } else {
+      createWindow(
+        searchButtonRef,
+        searchInputRef,
+        props.dockedTo,
+        windowState.bounds,
+        setResultsWindow
+      ).catch(err => console.error(err));
+    }
+  }, [closeResultsWindow, props.dockedTo, windowState.bounds]);
+
+  useEffect(() => {
+    const stockFluxSearch = async () => {
+      if (debouncedQuery) {
+        dispatch({ type: SEARCHING });
+        try {
+          const currentRequest = StockFlux.stockFluxSearch(debouncedQuery).then(
+            stockFluxResults => {
+              if (latestRequest === currentRequest) {
+                dispatch({ type: SUCCESS, results: stockFluxResults });
+              }
+            }
+          );
+          latestRequest = currentRequest;
+        } catch {
+          dispatch({ type: ERROR });
+        }
+      } else {
+        dispatch({ type: INITIALISE });
+      }
+    };
+    stockFluxSearch();
+  }, [debouncedQuery]);
+
+  useEffect(() => {
+    closeResultsWindow();
+  }, [closeResultsWindow, props.dockedTo]);
+
+  useEffect(() => {
+    const handler = setTimeout(
+      () => setDebouncedQuery(query),
+      SEARCH_TIMEOUT_INTERVAL
+    );
+    return () => clearTimeout(handler);
+  }, [query]);
+
+  useEffect(() => {
+    if (!isSearching && debouncedQuery !== null) {
+      if (results && results.length && debouncedQuery.length !== 0) {
+        const html = results.map(result => (
+          <SearchResult
+            key={result.code}
+            code={result.code}
+            name={result.name}
+          />
+        ));
+        populateResultsContainer(html, resultsWindow);
+      } else {
+        populateResultsContainer(
+          debouncedQuery.length === 0 ? INITIAL_MESSAGE_HTML : NO_MATCHES_HTML,
+          resultsWindow
+        );
+      }
+    }
+  }, [debouncedQuery, isSearching, props.dockedTo, results]);
+
+  window.fin.Window.getCurrentSync()
```

Comment:
Do we need to do this every render or can we do this more sparingly inside a useEffect?